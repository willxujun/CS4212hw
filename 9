//Source: Taglia vector_class.j
class Main123 {
	Void main () {
		return null;
	}
}

class Vector {
	Int x;
	Int y;

	Void addVector(Vector that) {
		this.x = this.x + that.x;
		this.y = this.y + that.y;
	}

	Void scaleVector(Int x) {
		this.x = this.x * x;
		this.y = this.y * x;
	}

	Int dotVector(Vector that) {
		Int result;
		result = result + this.x * that.x;
		result = result + this.y * that.y;
		return result;
	}

	Bool isEqual_v1(Vector that) {
		if (that == null) {
			return false;
		} else {
			return this.x == that.x && this.y == that.y;
		}
	}

	Bool isEqual_v2(Vector that) {
		if (!that) {
			return false;
		} else {
			return this.x == that.x && this.y == that.y;
		}
	}

	Bool isEqual_v3(Vector that) {
		return !that && this.x == that.x && this.y == that.y;
	}
}
