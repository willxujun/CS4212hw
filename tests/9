/* method overloading - error */
class Main {
    Void main() {
        A a;
        Int value;
        a = new A();
        value = a.getValue(1,2,2);
        value = a.getValue(3);
        value = a.getValue(4,true);
        return;
    }
}

class A {
    Int value;
    B b;
    
    B getB() {
        return b;
    }

    Int getValue(Int a, Int b, Int c) {
        this.value = -(a*b-c);
        return value;
    }
    Int getValue(Int a) {
        return a;
    }
    Int getValue(Int a, Bool b) {
        if(b) {
            return a;
        } else {
            return 0;
        }
    }
    
}

class B {
    C c;
    
    C getC() {
        return c;
    }
}

class C {
    A a;
    D d;

    A getA() {
        return a;
    }

    D getD() {
        return d;
    }
}

class D {
    A a;

    A getA() {
        return this.a;
    }
}
