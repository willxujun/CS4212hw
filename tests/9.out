Parse error, or failed distinct name check. See parse tree for detail.
Program
  Class Main
    VariableList
    MethodList
      Void main
        FormalList
        VariableList
          A a
          Int value
        StatementList
          Assign a
            New A
          Assign value
            Dispatch
              Access getValue
                (Id a)
              ExpressionList
                (Int 1)
                (Int 2)
                (Int 2)
          Assign value
            Dispatch
              Access getValue
                (Id a)
              ExpressionList
                (Int 3)
          Assign value
            Dispatch
              Access getValue
                (Id a)
              ExpressionList
                (Int 4)
                (Bool true)
          Return
  Error: Method cannot overload in method getValue of class A
  Class B
    VariableList
      C c
    MethodList
      C getC
        FormalList
        VariableList
        StatementList
          Return
            (Id c)
  Class C
    VariableList
      A a
      D d
    MethodList
      A getA
        FormalList
        VariableList
        StatementList
          Return
            (Id a)
      D getD
        FormalList
        VariableList
        StatementList
          Return
            (Id d)
  Class D
    VariableList
      A a
    MethodList
      A getA
        FormalList
        VariableList
        StatementList
          Return
            Access a
              This

